pallets/parachain-system/src/lib.rs

```diff
@@ -946,6 +948,84 @@ impl<T: Config> GetChannelInfo for Pallet<T> {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Updates inherent data to only contain messages that weren't already processed
+	/// by the runtime based on MQC heads.
+	///
+	/// # Panics
+	///
+	/// Panics if MQC heads don't match even after exhausting incoming message queue, either
+	/// downward or horizontal.
+	fn drop_processed_messages_from_inherent(para_inherent: &mut ParachainInherentData) {
+		let ParachainInherentData {
+			validation_data,
+			relay_chain_state,
+			downward_messages,
+			horizontal_messages,
+		} = para_inherent;
+
+		// Last MQC heads.
+		let last_dmq_head = <LastDmqMqcHead<T>>::get();
+		let last_hrmp_heads = <LastHrmpMqcHeads<T>>::get();
+
+		// Read MQC from relay chain state.
+		let relay_state_proof = RelayChainStateProof::new(
+			T::SelfParaId::get(),
+			validation_data.relay_parent_storage_root,
+			relay_chain_state.clone(),
+		)
+		.expect("Invalid relay chain state proof");
+		let relevant_messaging_state = relay_state_proof
+			.read_messaging_state_snapshot()
+			.expect("Invalid messaging state in relay chain state proof");
+
+		let relay_dmq_head = MessageQueueChain::from(relevant_messaging_state.dmq_mqc_head);
+		let relay_hrmp_heads =
+			relevant_messaging_state.ingress_channels.iter().map(|(para_id, channel)| {
+				let mqc_head = channel.mqc_head.unwrap_or_default();
+				(para_id, MessageQueueChain::from(mqc_head))
+			});
+
+		// DMQ.
+		let dmq_processed_num = (0..=downward_messages.len())
+			.find(|processed| {
```

Comment:
this algorithm seems very weird.

Can we not just do a single loop over downward messages with a helper function like this:

```rust
let mut current_mqc_head = last_dmq_head;
let target_mqc_head = relay_dmq_head;

for (i, messages) in 0..messages.iter().enumerate() {
    if current_mqc_head = target_mqc_head { return i }
    current_mqc_head.extend_downwards(messages[i]);
}

panic!("messages in inherent are not consistent with MQC");
```

Reply:
As long as we don't know how many messages were already processed, I don't think O(N^2) can be avoided.

In your example, suppose `current_mqc_head` points to somewhere in the middle of the queue, it will certainly panic.

Reply:
I don't think `LastRelayChainBlockNumber` is guaranteed to change?

We can use small optimization and compare last mqc with relay mqc, assuming the whole queue was exhausted previously, otherwise, perform search.

Reply:
> I don't think LastRelayChainBlockNumber is guaranteed to change?

It feels safe to rely on this being updated every block, if that's what you mean.

If the `LastRelayChainBlockNumber = N`, then we can safely omit all messages `sent_at <= N` as they've presumably already been processed. The MQC head and pending messages are drawn out of a child relay-chain block M, so we'd end up providing all messages where `N < sent_at <= M`. If the current relay-parent is the same as the previous, then N = M and no messages will have `N < sent_at <= N` and we'd provide an empty list, which is the desired behavior.